diff -up jss-4.2.6/mozilla/security/jss/org/mozilla/jss/CryptoManager.java.cfu jss-4.2.6/mozilla/security/jss/org/mozilla/jss/CryptoManager.java
--- jss-4.2.6/mozilla/security/jss/org/mozilla/jss/CryptoManager.java.cfu	2015-07-08 12:02:13.192000000 +0200
+++ jss-4.2.6/mozilla/security/jss/org/mozilla/jss/CryptoManager.java	2015-07-08 12:04:56.213000000 +0200
@@ -1404,17 +1404,23 @@ public final class CryptoManager impleme
     synchronized static void loadNativeLibraries()
     {
         if( ! mNativeLibrariesLoaded ) {
-            try {
+            try { // 64 bit rhel/fedora
                 System.load( "/usr/lib64/jss/libjss4.so" );
                 Debug.trace(Debug.VERBOSE, "jss library loaded");
                 mNativeLibrariesLoaded = true;
             } catch( UnsatisfiedLinkError e ) {
-                try {
+                try { // 32 bit rhel/fedora
                     System.load( "/usr/lib/jss/libjss4.so" );
                     Debug.trace(Debug.VERBOSE, "jss library loaded");
                     mNativeLibrariesLoaded = true;
                 } catch( UnsatisfiedLinkError f ) {
-                    Debug.trace(Debug.VERBOSE, "jss library load failed");
+                    try {// possibly other platforms
+                        System.loadLibrary( "jss4" );
+                        Debug.trace(Debug.VERBOSE, "jss library loaded");
+                        mNativeLibrariesLoaded = true;
+                    } catch( UnsatisfiedLinkError g ) {
+                        Debug.trace(Debug.VERBOSE, "jss library load failed");
+                    }
                 }
             }
         }
